File: problems/148/solution.go
package main
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func sortList(head *ListNode) *ListNode {
    if head == nil || head.Next == nil {
        return head
    }
    
    mid := findMid(head)
    left, right := sortList(head), sortList(mid)
    
    return merge(left, right)
}

func merge(left, right *ListNode) *ListNode {
    dummy := &ListNode{}
    tail := dummy
    
    for left != nil && right != nil {
        if left.Val < right.Val {
            tail.Next = left
            left = left.Next
        } else {
            tail.Next = right
            right = right.Next
        }
        
        tail = tail.Next
    }
    
    if left != nil {
        tail.Next = left
    } else {
        tail.Next = right
    }
    
    return dummy.Next
}

func findMid(node *ListNode) *ListNode {
    var midPrev *ListNode = nil
    
    for node != nil && node.Next != nil {
        if (midPrev != nil) {
            midPrev = midPrev.Next
        } else {
            midPrev = node
        }
        
        node = node.Next.Next
    }
    
    mid := midPrev.Next
    midPrev.Next = nil
    
    return mid
}

File: problems/148/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* sortList(ListNode* head) {
        if (!head || !head->next) {
            return head;
        }
        
        ListNode* mid = findMid(head);
        ListNode* left = sortList(head);
        ListNode* right = sortList(mid);
        
        return merge(left, right);
    }
    
    ListNode* merge(ListNode* left, ListNode* right) {
        ListNode dummy(0);
        ListNode* tail = &dummy;
        
        while (left && right) {
            if (left->val < right->val) {
                tail->next = left;
                left = left->next;
            } else {
                tail->next = right;
                right = right->next;
            }
            
            tail = tail->next;
        }
        
        tail->next = left ? left : right;
        
        return dummy.next;
    }
    
    ListNode* findMid(ListNode* node) {
        ListNode* midPrev = nullptr;
        
        while (node && node->next) {
            midPrev = midPrev ? midPrev->next : node;
            node = node->next->next;
        }
        
        ListNode* mid = midPrev->next;
        midPrev->next = nullptr;
        
        return mid;
    }
};